File: bot.py
import json
import os
import requests
import base64
from datetime import date
from dotenv import load_dotenv
from telegram import Update
from telegram.ext import ApplicationBuilder, CommandHandler, MessageHandler, ContextTypes, filters

# ========================
# CONFIGURAÇÃO
# ========================
load_dotenv()
TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")

if not TOKEN or not GEMINI_API_KEY:
    raise RuntimeError("❌ Verifique se TELEGRAM_BOT_TOKEN e GEMINI_API_KEY estão no seu arquivo .env")

DATA_FILE = "data.json"
META_CALORIAS = 3300

# ========================
# UTILITÁRIOS
# ========================
def load_data():
    if not os.path.exists(DATA_FILE): return {}
    try:
        with open(DATA_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except: return {}

def save_data(data):
    with open(DATA_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

def encode_image(image_path):
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode('utf-8')

# ========================
# INTEGRAÇÃO GEMINI 1.5 FLASH
# ========================
def ask_gemini(description=None, image_path=None):
    # Endpoint v1 oficial (mais estável que v1beta)
    url = f"https://generativelanguage.googleapis.com/v1/models/gemini-1.5-flash:generateContent?key={GEMINI_API_KEY}"
    
    prompt = """Você é um nutricionista. Analise o alimento e retorne APENAS um JSON.
    Formato: {"food": "nome", "calories": 000}
    Se não for comida, use null nos campos."""

    parts = [{"text": prompt}]
    
    if description:
        parts.append({"text": f"Usuário descreveu: {description}"})
    
    if image_path:
        parts.append({
            "inline_data": {
                "mime_type": "image/jpeg",
                "data": encode_image(image_path)
            }
        })

    payload = {"contents": [{"parts": parts}]}

    try:
        r = requests.post(url, json=payload, timeout=30)
        
        if r.status_code == 404:
            print("❌ ERRO 404: O modelo Gemini 1.5 Flash não foi encontrado. Verifique sua chave de API.")
            return None
        if r.status_code == 400:
            print("❌ ERRO 400: Requisição inválida. Verifique o formato dos dados.")
            return None

        r.raise_for_status()
        res_data = r.json()
        
        # Extrai e limpa o JSON da resposta
        raw_text = res_data["candidates"][0]["content"]["parts"][0]["text"].strip()
        clean_json = raw_text.replace("```json", "").replace("```", "").strip()
        
        return json.loads(clean_json)
    except Exception as e:
        print(f"⚠️ Erro na API Gemini: {e}")
        return None

# ========================
# HANDLERS DO TELEGRAM
# ========================
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text("🍎 Envie uma foto do prato ou texto para contar calorias!")

async def handle_input(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = str(update.message.from_user.id)
    today = str(date.today())
    image_path = None

    status_msg = await update.message.reply_text("⏳ Analisando...")

    if update.message.photo:
        photo_file = await update.message.photo[-1].get_file()
        image_path = f"temp_{user_id}.jpg"
        await photo_file.download_to_drive(image_path)
        result = ask_gemini(image_path=image_path)
    else:
        result = ask_gemini(description=update.message.text)

    if result and result.get("food"):
        data = load_data()
        data.setdefault(user_id, {}).setdefault(today, {"calories": 0})
        
        cal = result["calories"]
        data[user_id][today]["calories"] += cal
        save_data(data)

        await status_msg.edit_text(
            f"✅ {result['food']}\n🔥 +{cal} kcal\n📊 Total hoje: {data[user_id][today]['calories']} kcal"
        )
    else:
        await status_msg.edit_text("❌ Não reconheci o alimento. Tente descrever melhor.")

    if image_path and os.path.exists(image_path):
        os.remove(image_path)

# ========================
# EXECUÇÃO
# ========================
if __name__ == "__main__":
    print("🚀 Bot Iniciado!")
    app = ApplicationBuilder().token(TOKEN).build()
    app.add_handler(CommandHandler("start", start))
    app.add_handler(MessageHandler(filters.TEXT | filters.PHOTO & ~filters.COMMAND, handle_input))
    app.run_polling()

File: bot.rb
require "telegram/bot"
require "net/http"
require "json"
require "bigdecimal"
require "bigdecimal/util"
require "date"
require "base64"
require "logger"

# =========================
# 🔑 TOKENS (COLE OS SEUS)
# =========================
TOKEN = "8460032402:AAH1-9x5GpyD30I_bBx6IjDAqFIp_2FV5Zo"
GEMINI_KEY = "AIzaSyBQbjjMx_5sQ4bNlfkJ5NFTpAmKVYvCOYc"
USDA_KEY = "WjwD1SJlqgaJfVWee01JeTkTZF8Cx3e9ShnTIhhH"

# =========================
# 🎯 METAS DIÁRIAS
# =========================
META = {
  calories: 3300.to_d,
  protein:  175.to_d,
  fat:      95.to_d,
  carbs:    435.to_d
}

DATA_FILE = "data.json"

# =========================
# 📦 STORAGE
# =========================
def load_data
  File.exist?(DATA_FILE) ? JSON.parse(File.read(DATA_FILE)) : {}
end

def save_data(data)
  File.write(DATA_FILE, JSON.pretty_generate(data))
end

# =========================
# 🥗 USDA
# =========================
def fetch_usda(food)
  uri = URI(
    "https://api.nal.usda.gov/fdc/v1/foods/search" \
    "?api_key=#{USDA_KEY}&query=#{URI.encode_www_form_component(food)}&pageSize=1"
  )

  res = JSON.parse(Net::HTTP.get(uri))
  nutrients = res.dig("foods", 0, "foodNutrients") || []

  get = ->(name) {
    n = nutrients.find { |x| x["nutrientName"].to_s.downcase.include?(name) }
    n ? n["value"].to_d : 0.to_d
  }

  {
    calories: get.call("energy"),
    protein:  get.call("protein"),
    fat:      get.call("fat"),
    carbs:    get.call("carbohydrate")
  }
end

# =========================
# 🤖 GEMINI (VISÃO)
# =========================
def identify_food(text, image_path)
  image_base64 = Base64.strict_encode64(File.binread(image_path))

  prompt = <<~PROMPT
    Analise a imagem e o texto "#{text}".
    Retorne APENAS JSON neste formato:
    { "food": "nome", "grams": numero }
  PROMPT

  body = {
    contents: [
      {
        parts: [
          { text: prompt },
          {
            inline_data: {
              mime_type: "image/jpeg",
              data: image_base64
            }
          }
        ]
      }
    ]
  }

  uri = URI(
    "https://generativelanguage.googleapis.com/v1beta/" \
    "models/gemini-pro-vision:generateContent?key=#{GEMINI_KEY}"
  )

  res = Net::HTTP.post(uri, body.to_json, "Content-Type" => "application/json")
  json = JSON.parse(res.body)

  JSON.parse(json["candidates"][0]["content"]["parts"][0]["text"])
end

# =========================
# 🚨 FORÇA POLLING (REMOVE WEBHOOK)
# =========================
puts "🧹 Removendo webhook (garantindo polling)..."
Net::HTTP.get(URI("https://api.telegram.org/bot#{TOKEN}/deleteWebhook"))

puts "🤖 Bot rodando... pressione CTRL+C para parar"

# =========================
# 📲 TELEGRAM BOT
# =========================
Telegram::Bot::Client.run(TOKEN, logger: Logger.new($stdout)) do |bot|
  data = load_data
  pending = {}

  bot.listen do |msg|
    next unless msg.from

    user = msg.from.id.to_s
    today = Date.today.to_s

    data[user] ||= {}
    data[user][today] ||= META.transform_values { 0.to_d }

    puts "📩 Mensagem recebida: #{msg.text || '[foto]'}"

    if msg.text == "/start"
      bot.api.send_message(
        chat_id: msg.chat.id,
        text: "📸 Envie a foto da refeição + descrição\n" \
              "'primeira refeição' → novo dia\n" \
              "/resumo → média semanal"
      )
    end

    if msg.text == "primeira refeição"
      data[user][today] = META.transform_values { 0.to_d }
      save_data(data)
      bot.api.send_message(chat_id: msg.chat.id, text: "🔄 Novo dia iniciado")
    end

    if msg.photo
      file = bot.api.get_file(file_id: msg.photo.last.file_id)
      path = "tmp_#{user}.jpg"
      File.write(
        path,
        Net::HTTP.get(
          URI("https://api.telegram.org/file/bot#{TOKEN}/#{file["result"]["file_path"]}")
        )
      )
      pending[user] = { image: path }
      bot.api.send_message(chat_id: msg.chat.id, text: "📸 Foto recebida! Agora descreva.")
    end

    if pending[user]&.dig(:image) && msg.text && !msg.text.start_with?("/")
      info = identify_food(msg.text, pending[user][:image])
      base = fetch_usda(info["food"])

      pending[user] = {
        grams: info["grams"].to_d,
        base: base
      }

      bot.api.send_message(
        chat_id: msg.chat.id,
        text: "🍽️ #{info["food"]}\n" \
              "📏 Estimado: #{info["grams"]}g\n" \
              "Digite a quantidade real ou 'ok'"
      )
    end

    if pending[user]&.dig(:base) && msg.text
      grams = msg.text == "ok" ? pending[user][:grams] : msg.text.to_d
      factor = grams / 100

      META.each_key do |k|
        data[user][today][k] += pending[user][:base][k] * factor
      end

      pending.delete(user)
      save_data(data)

      c = data[user][today]
      rest = META[:calories] - c["calories"]

      bot.api.send_message(
        chat_id: msg.chat.id,
        text: "🔥 #{c["calories"].to_i}/3300 kcal\n" \
              "🥩 #{c["protein"].to_i}/175g\n" \
              "🥑 #{c["fat"].to_i}/95g\n" \
              "🍞 #{c["carbs"].to_i}/435g\n\n" \
              "#{rest > 0 ? "👉 Restam #{rest.to_i} kcal" : "⚠️ Meta ultrapassada"}"
      )
    end

    if msg.text == "/resumo"
      days = (0..6).map { |i| (Date.today - i).to_s }
      week = days.map { |d| data[user][d] }.compact

      avg = META.transform_values { 0.to_d }
      week.each { |d| avg.each_key { |k| avg[k] += d[k].to_d } }
      avg.each_key { |k| avg[k] /= week.size if week.any? }

      bot.api.send_message(
        chat_id: msg.chat.id,
        text: "📊 Média últimos 7 dias\n" \
              "🔥 #{avg[:calories].to_i} kcal\n" \
              "🥩 #{avg[:protein].to_i}g\n" \
              "🥑 #{avg[:fat].to_i}g\n" \
              "🍞 #{avg[:carbs].to_i}g"
      )
    end
  end
end
